src/awkward/_v2/cpp-headers/awkward/LayoutBuilder.h

```diff
@@ -0,0 +1,1706 @@
+// BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
+
+#ifndef AWKWARD_LAYOUTBUILDER_H_
+#define AWKWARD_LAYOUTBUILDER_H_
+
+#include "awkward/GrowableBuffer.h"
+#include "awkward/utils.h"
+
+#include <map>
+#include <algorithm>
+#include <tuple>
+#include <string>
+
+namespace awkward {
+
+  namespace LayoutBuilder {
+
+  template<std::size_t ENUM, typename BUILDER>
+  class Field {
+  public:
+    using Builder = BUILDER;
+
+    std::string index_as_field() const {
+      return std::to_string(static_cast<int>(index));
+    }
+
+    template<typename PRIMITIVE>
+    void
+    append(PRIMITIVE x) {
+      builder.append(x);
+    }
+
+    const std::size_t index = ENUM;
+    Builder builder;
+  };
+
+  template <unsigned INITIAL, typename PRIMITIVE>
```

Comment:
Why is the initial size of the buffer a compile-time constant?

This should become a runtime value, passed down through instance constructors, not template parameters.

With it in the template parameter, every C++ user is going to have to specify `INITIAL` at every node of the tree. That's a lot of boilerplate for a value they usually don't care about. They'll usually take a default that gets passed in from above, or through a `BuilderOptions`.

The reason it's a parameter at all, instead of hard-coded into the code, is so that we have some leverage when doing performance tests. Users, even C++  users, will trust whatever optimal value we come up with from those performance tests. They're not going to want to have to type it.

Reply:
The users will not see it - we can type alias it as in the test examples. The reason why we need it as a template parameter - we use the default constructors. The builders are allocated on the stack and so their GrowableBuffers. We could change it, but let’s discuss it on Wednesday first. Thanks

Reply:
Okay, the initial GrowableBuffer size has to be a compile-time constant for a C++ reason: you need to use default constructors. (Why is that?)

And if GrowableBuffers are parameterized by more than one parameter—if they need to have a different first panel size from subsequent panel sizes as we discussed in the last meeting—do all of those have to become template parameters?

If this is the case, then we need to provide type aliases as part of the header-only package, with some choice of GrowableBuffer parameters hard-coded. C++ users of this package need to be able to say `LayoutBuilder::Numpy<float>` without having to set up an alias on their own.